src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3591,6 +3607,15 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                   case _ => tp
                 }
 
+                // Inline RHS of ValDef for right-associative by-value operator desugaring
+                val args2 = (args1, mt.params) match {
+                  case ((ident: Ident) :: Nil, param :: Nil) if param.isByNameParam && rightAssocValDefs.contains(ident.symbol) =>
+                    inlinedRightAssocValDefs += ident.symbol
+                    val rhs = rightAssocValDefs(ident.symbol)
```

Comment:
I'd prefer a symbol attachment over the two collections. Something like `class RightAssocValDefAttachment(var rhsInlined)`, added in `typedValDef`, the var could be set to `true` here. Then you can use `getAndRemoveAttachment` in `typedBlock`.

We could even just add an empty marker attachment `RightAssocValDefInlined` here, and skip the test in `typedValDef` - what value does it add?